lib/mongoose_push/service/apns/supervisor.ex

```diff
@@ -31,45 +31,104 @@ defmodule MongoosePush.Service.APNS.Supervisor do
   end
 
   defp create_sparrow_config(pool_configs) do
-    dev_pools =
+    dev_cert_pools =
       pool_configs
-      |> Enum.filter(fn {_, pool_config} -> pool_config[:mode] == :dev end)
-      |> List.foldl([], &convert_pool_to_sparrow/2)
+      |> Enum.filter(fn {_, pool_config} ->
+        pool_config[:mode] == :dev and
+          pool_config[:auth][:type] == :certificate_based
+      end)
+      |> Enum.map(&convert_cert_pool_to_sparrow/1)
 
-    prod_pools =
+    prod_cert_pools =
       pool_configs
-      |> Enum.filter(fn {_, pool_config} -> pool_config[:mode] == :prod end)
-      |> List.foldl([], &convert_pool_to_sparrow/2)
+      |> Enum.filter(fn {_, pool_config} ->
+        pool_config[:mode] == :prod and
+          pool_config[:auth][:type] == :certificate_based
+      end)
+      |> Enum.map(&convert_cert_pool_to_sparrow/1)
 
-    [{:dev, dev_pools}, {:prod, prod_pools}]
+    {dev_token_pools, dev_tokens} =
+      pool_configs
+      |> Enum.filter(fn {_, pool_config} ->
+        pool_config[:mode] == :dev and
+          pool_config[:auth][:type] == :token_based
+      end)
+      |> List.foldl({[], []}, &convert_token_pool_to_sparrow/2)
+
+    {prod_token_pools, prod_tokens} =
+      pool_configs
+      |> Enum.filter(fn {_, pool_config} ->
+        pool_config[:mode] == :prod and
+          pool_config[:auth][:type] == :token_based
+      end)
+      |> List.foldl({[], []}, &convert_token_pool_to_sparrow/2)
+
+    [
+      {:dev, dev_cert_pools ++ dev_token_pools},
+      {:prod, prod_cert_pools ++ prod_token_pools},
+      {:tokens, dev_tokens ++ prod_tokens}
+    ]
   end
 
-  defp convert_pool_to_sparrow({pool_name, pool_config}, mode_list) do
+  defp convert_cert_pool_to_sparrow({pool_name, pool_config}) do
     auth_type = {:auth_type, :certificate_based}
-    cert = {:cert, pool_config[:cert]}
-    key = {:key, pool_config[:key]}
+    cert = {:cert, pool_config[:auth][:cert]}
+    key = {:key, pool_config[:auth][:key]}
     pool_size = {:worker_num, pool_config[:pool_size]}
 
     endpoint_mode = @default_endpoints[pool_config[:mode]]
     endpoint = {:endpoint, pool_config[:endpoint] || endpoint_mode}
 
-    port_config =
-      case pool_config[:use_2197] do
-        true ->
-          2197
+    port_config = if pool_config[:use_2197], do: 2197, else: nil
+    port = {:port, port_config}
+
+    name = {:pool_name, pool_name}
+
+    [auth_type, cert, key, pool_size, endpoint, port, name]
+    |> Enum.filter(fn {_key, value} -> !is_nil(value) end)
+  end
+
+  defp convert_token_pool_to_sparrow({pool_name, pool_config}, {mode_list, tokens}) do
```

Comment:
There seem to be a lot of code duplication between those to "convert" functions. Can we extract common parts to another, and keep only auth parts separate?